resource_alks_iamrole.go

```diff
@@ -192,56 +184,39 @@ func resourceAlksIamTrustRoleCreate(d *schema.ResourceData, meta interface{}) er
 
 	log.Printf("[INFO] alks_iamtrustrole.id: %v", d.Id())
 
-	return resourceAlksIamRoleRead(d, meta)
+	return diags
 }
 
-func resourceAlksIamRoleDelete(d *schema.ResourceData, meta interface{}) error {
+func resourceAlksIamRoleDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	log.Printf("[INFO] ALKS IAM Role Delete")
 
 	client := meta.(*alks.Client)
 	if err := validateIAMEnabled(client); err != nil {
-		return err
+		return diag.FromErr(err)
 	}
 
 	if err := client.DeleteIamRole(d.Id()); err != nil {
-		return err
+		return diag.FromErr(err)
 	}
 
 	return nil
 }
 
-func resourceAlksIamRoleExists(d *schema.ResourceData, meta interface{}) (b bool, e error) {
-	log.Printf("[INFO] ALKS IAM Role Exists")
+func resourceAlksIamRoleRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	log.Printf("[INFO] ALKS IAM Role Read")
 
 	client := meta.(*alks.Client)
 
-	foundRole, err := client.GetIamRole(d.Id())
-
-	if err != nil {
-		// TODO: Clean-up this logic, likely by improving the error responses from `alks-go`
-		if strings.Contains(err.Error(), "Role not found") {
-			return false, nil
-		}
-
-		return false, err
-	}
-
-	if foundRole == nil {
-		return false, nil
+	// Check if role exists.
+	if d.Id() == "" || d.Id() == "none" {
+		return nil
 	}
 
-	return true, nil
-}
-
-func resourceAlksIamRoleRead(d *schema.ResourceData, meta interface{}) error {
-	log.Printf("[INFO] ALKS IAM Role Read")
-
-	client := meta.(*alks.Client)
 	foundRole, err := client.GetIamRole(d.Id())
 
 	if err != nil {
 		d.SetId("")
-		return err
+		return nil
```

Comment:
Feel free to tell me I'm wrong because I'm just noticing a pattern, but shouldn't this be `diag.FromErr(err)`

Reply:
Okay so these are quirks with Terraform itself and the behavior we expect to return the user. We _always_ return nil on a read operation whenever a resource is not found - hence why the Id is set to blank. This is just the behavior Terraform wants when we check to see if the resource exists, and on that line - it does not exist, so we return nil NOT an error. Since this is a read operation btw or a GET call on the API, we only expect a yes or a no, there shouldn't be a maybe because Terraform doesn't care, so we return nil. I hope that explains this part - it took me a few reads to get it.